model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanAuthenticationSessionProviderFactory.java

```diff
@@ -94,7 +96,7 @@ public int order() {
     }
 
     private AuthenticationSessionTransaction createAndEnlistTransaction(KeycloakSession session) {
-        var tx = new AuthenticationSessionTransaction(cache);
+        var tx = new AuthenticationSessionTransaction(cacheHolder.cache(), new ByRealmIdQueryConditionalRemover<>(PROTO_ENTITY, cacheHolder.executor()));
```

Comment:
The idea is to have a single implementation per entity, see [1]. In this case, I'm using the same class for both entities since they are the same.

The implementation will handle all the fields that can be used for removals. 

[1] https://github.com/pruivo/keycloak/blob/2cd8e366767d55904b9722402a00c9a715dfb2e8/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/remote/remover/query/ClientSessionQueryConditionalRemover.java

Reply:
Ok, that looks good to me! Thanks for sharing. 

Last question, I see that for get queries, you used a different approach than to removals (something like named queries we have for JPA): https://github.com/pruivo/keycloak/blob/2cd8e366767d55904b9722402a00c9a715dfb2e8/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/query/ClientSessionQueries.java#L33-L36

Have you considered using the same approach as for removers to have more flexibility in building clauses? I am not saying it is wrong, the advantage is that it is similar to JPA but I am curious what was the reason to have different approach than for removers.

Reply:
I did not. I was trying to achieve a single delete statement with a dynamic condition. Taking the client session as an example, if you remove by realm id and by user session id in the same Keycloak transaction, it will create the condition `(realmId = :p0) || (userSessionId = :p1)`

If I had understood what you suggested, I would have 2 delete statements:

`delete from <client session> where realmId = :p0` and `delete from <client session> where userSessionId = :p0`

The disadvantage of having 2 delete statements running concurrently is, that if a key matches both conditions, you would have 2 removals concurrently happen in the cluster (and sent to the other site).

Reply:
Yes, I understand that. I did not suggest using two queries. What I am saying is that we have different approaches for fetch queries and removes. While we have dynamic WHERE clauses creation for removes, in the commit you shared, we don't have that for fetching and we have static named queries: https://github.com/pruivo/keycloak/blob/2cd8e366767d55904b9722402a00c9a715dfb2e8/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/query/ClientSessionQueries.java#L33-L36

I was wondering why there is a different approach and whether in future we could benefit from the dynamic WHERE clauses creation also for fetching.

But this is not relevant to this PR so I think we can discuss this later if needed.